### PR TITLE
IDBNext, implementing find natively #6543

### PR DESCRIPTION
**Pull request by SCdF**
Here is a hopefully mergable chunk of code around implementing find natively. It's a lot of code, sorry!

It's at a point where I would be happy for this to be merged into the main codebase, though not to become the default adapter quite yet.

We still have more work to do before that (eg there is no migration path that I'm aware of). I'm interested in what you think about instead exposing this as an optional adapter so people can use it / test it if they want, e.g.:

```js
const PouchDB = require('pouchdb-core');
PouchDB.plugin(require('pouchdb-adapter-indexeddb'));
```

It could then be moved later to the default adapter once we have a migration path.

Alternatively, we can just leave it hidden behind NEXT=1 until we're super duper ready to go.

## Additional note: Write cloning

To maintain as much support with CouchDB as possible we're forced to clone data on write and modify it so that IndexedDB elements it would otherwise be unable to.

Specifically:
 -  `true`, `false` and `null` are not indexable by IndexedDB, so are instead rewritten as very small integers (which also maintains them in the correct type ordering that CouchDB expects)
 -  IndexedDB only supports indexes against paths that are composed of valid JS variable names. CouchDB supports anything that JSON supports. We rewrite valid JSON but invalid JS into valid JS names.

This does affect document write performance, but only for large, deep docs. This could almost certainly improved by improving the clone algorithm (I do not profess to be a JS performance expert, and I'm sure it using recursive function calls gives us unhelpful stack thrashing).

| | Existing  | Idbnext | Idbnext+rewrite |
|--- | --- | --- | --- |
| basic-inserts | 3.10 | 2.70 | 2.70 |
| basic-inserts-large-docs | 3.20 | 3.10 | 3.10 |
| basic-inserts-deep-docs | 9.2 | 8.3 | 12.20 |

I will look at tuning it later. If that goes nowhere and we are worried about it, we could make it non default and hide it behind a CouchDB compatibility flag:

```js
new PouchDB('foo', {couchdb_compat: true});
```

With documentation on what running without compatibility limits you to.

Let me know what you think.

## Additional note: General performance

Running the perf tests on a 2015 13" MBP on Chrome 70 I get the following results between idb and indexeddb (idbnext), fasted time bolded.

| | indexeddb (new) | idb (old) |
| --- | --- | --- |
| basic-inserts | **2.60** | 3.30 |
| basic-inserts-large-docs | **3.00** | 3.30 |
| basic-inserts-deep-docs | 11.90 | **8.95** |
| bulk-inserts | **29.05** | 41.90 |
| basic-updates | **81.20** | 81.55 |
| basic-gets | **0.50** | 0.70 |
| all-docs-skip-limit | 272.30 | **92.25** |
| all-docs-startkey-endkey | 26.70 | **16.45** |
| all-docs-keys | **5.90** | 10.85 |
| all-docs-include-docs | **15.60** | 17.70 |
| pull-replication-one-generation | **2345.80** | 2462.00 |
| pull-replication-two-generation | **5795.20** | 6331.50 |
| temp-views | 5006.20 | **4959.60** |
| build-secondary-index | **693.90** | 798.20 |
| persisted-views | **20.90** | 25.00 |
| persisted-views-stale-ok | 17.85 | **17.50** |
| create-index | **991.30** | 8104.70 |
| simple-find-query | **3.05** | 8.50 |
| simple-find-query-no-index | **577.85** | 1154.35 |
| complex-find-query | **657.75** | 3267.90 |
| complex-find-query-no-index | **614.85** | 1115.45 |
| multi-field-query | **673.50** | 3328.45 |
| basic-attachments | 12.10 | **8.70** |

So we're mostly better, and _vastly_ better when it comes to find (`create-index` scales continuously better the more docs you create to be indexed, in this benchmark it's only 1000).

There are definitely a few places here worth investigating though as separate pieces of work.

In general I think it's worth going through complicated api calls that have a lot of generic code in front of them, because it looks like there is room for improvement there.

For example, bulk get with attachments seems to be:
 - Generic code does a loop over id, and for each:
   - Performs an adapter-level _get call (+1 transaction opened)
   - If {attachments: true} and there are attachments, for every attachment
     - Performs an adapter-level _getAttachment call (+1 transaction opened)

If you get 10 document with 3 attachments that's 40 opened transactions and 40 idb get interactions, where it really should be 1 transaction and 10 idb get interactions.

cc. @garrensmith, @daleharvey

### CODE REVIEW

File: packages/node_modules/pouchdb-adapter-indexeddb/src/find.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+// But IDBKeyRange is not a fan of shenanigans, so I've just gone with 12 layers
+// because it looks nice and surely that's enough!
+var IDB_COLLATE_LO = Number.NEGATIVE_INFINITY;
+var IDB_COLLATE_HI = [[[[[[[[[[[[]]]]]]]]]]]];
+
+//
+// TODO: this should be made offical somewhere and used by AllDocs / get /
+// changes etc as well.
+//
+function externaliseRecord(idbDoc) {
```
Comment (garrensmith) [inline packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: Do you still want to make this official and make it the only function we use to externalise?

File: packages/node_modules/pouchdb-adapter-indexeddb/src/find.js [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+    var fixed = [0].concat(key);
+
+    // REVIEWER: Read this block carefully.
+    //
+    // This code is replacing *all* instances both the upper and lower collate
+    // mark, regardless of where they are in the array, for non exact (eg
+    // ranged) queries.
+    //
+    // I am not sure if this is right.
+    fixed = fixed.map(function (k) {
```
Comment (garrensmith) [inline packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: I think you can just return here.

Comment (garrensmith) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: I don't think fix is a good name for the function. `Normalise` would be better.

Comment (garrensmith) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: Or something related to the fact we changing the keys to work with the way we store the data.

File: packages/node_modules/pouchdb-adapter-indexeddb/src/find.js (hunk 1 above)
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: It's on my list of TODOs to look into at some point yeah. It would just be a minor cleanup, but it would also help to double-check, document and standardise that we shouldn't use `doc.data` when returning docs to callers.

File: packages/node_modules/pouchdb-adapter-indexeddb/src/find.js (hunk 2 above)
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: `translate`?

File: packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+var IDB_NULL = Number.MIN_SAFE_INTEGER;
+var IDB_FALSE = Number.MIN_SAFE_INTEGER + 1;
+var IDB_TRUE = Number.MIN_SAFE_INTEGER + 2;
+
+var ALLOWABLE_SINGLE_FIELD = ['_', '$'];
+var ALLOWABLE = ALLOWABLE_SINGLE_FIELD.concat(['.']);
+
+// IndexedDB only allows valid JS names in its index paths, whereas JSON allows
```
[the remark this answers is not in the document]
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js]: yep will do

File: packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+// There are more complete solutions available, such as using:
+//    https://mothereff.in/js-variables
+// As a base to properly validate, but it's very complicated and would only be
+// worth looking into if act of rewriting slowed things down and we wanted to
+// reduce the instances in which the `bad` var below is hit.
+//
+// This can almost certainly be performance tuned.
+function sanitise(name, singleField) {
+  var sanitised = '';
+  var last = 0;
```
[the remark this answers is not in the document]
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js]: Good catch, yeah if anything it's `start` not `last`. For now I've changed it to `substringStart`, but I want to rework this area.

File: packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    var c = name[i];
+    var bad = nextBad;
+    nextBad = false;
+
+    if (!bad) {
+      if (!singleField && c === '\\') {
+        nextBad = true;
+      } else if (i === 0 && c >= '0' && c <= '9') {
+        bad = true;
+      } else if (!((c >= '0' && c <= '9') ||
```
[the remark this answers is not in the document]
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js]: Yeah agreed.

I'm going to write some performance tests specifically for this function, as well as some functional unit tests.

(both of these things are already exercised in performance and int. tests, but I'd like specific ones without external noise for here)

Once I have those done we can work out a balance of performance to readability

File: packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js [hunk 6]
```diff
@@ -2,73 +2,196 @@
...
-function createSchema(db) {
+    var expectedIndexes = {};
+
+    // NB: the only thing we're supporting here is the declared indexing
+    // fields nothing more.
+    results.filter(function (r) {
+      return r.deleted === 0;
+    }).map(function (r) {
+      return r.data;
+    }).forEach(function (ddoc) {
```
[the remark this answers is not in the document]
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js]: Yep true, agree 100%. I'll rewrite

File: packages/node_modules/pouchdb-adapter-indexeddb/src/util.js [hunk 7]
```diff
@@ -36,22 +37,59 @@ function processAttachment(name, src, doc, isBinary) {
...
-      error: err
-    };
-  }
+function rawIndexFields(ddoc, viewName) {
+  // fields are an array of either the string name of the field, or a key value
+  var fields = ddoc.views[viewName].options &&
+               ddoc.views[viewName].options.def &&
+               ddoc.views[viewName].options.def.fields || [];
+
+  // Either ['foo'] or [{'foo': 'desc'}]
```
[the remark this answers is not in the document]
Comment (garrensmith) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/util.js]: Yes. CouchDB makes sure that everything we work with is normalised and we can expect it in one way.

**Comment by SCdF** [comment]
Are you comfortable with us potentially breaking backwards compatibility when we migrate existing idb databases to idbnext?

(or I suppose we could also normalise data in that same migration)

File: tests/find/test-suite-1/test.zzz-suite-1.js [hunk 8]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-['http', 'local'].forEach(function (adapter) {
+['local', 'http'].forEach(function (adapter) {
```
[the remark this answers is not in the document]
Comment (SCdF) [reply tests/find/test-suite-1/test.zzz-suite-1.js]: Sorry this shouldn't have made it through, I'll revert.

I switched it because my confidence in local passing is lower than my confidence in http passing, so this way failures in my code came faster.

File: packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/rewrite.js]: I've simplified this, replacing my custom work with regex. You can see my investigation here: https://github.com/SCdF/rewrite-perf

tl;dr; it's slower but not that much slower, and when you add the IO and other parts of PouchDB around it, it barely matters:
```
BEFORE:
Starting test: basic-inserts with 1 assertions and 1000 iterations... median: 2.300000051036477 ms
Starting test: basic-inserts-large-docs with 1 assertions and 100 iterations... median: 3.000000026077032 ms
Starting test: basic-inserts-deep-docs with 1 assertions and 100 iterations... median: 12.049999961163849 ms

AFTER:
Starting test: basic-inserts with 1 assertions and 1000 iterations... median: 2.199999988079071 ms
Starting test: basic-inserts-large-docs with 1 assertions and 100 iterations... median: 3.000000026077032 ms
Starting test: basic-inserts-deep-docs with 1 assertions and 100 iterations... median: 12.64999998966232 ms
```

File: tests/find/test-suite-1/test.data-type-order.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+      });
+    });
+
+    // IndexedDB does not index obejcts, so this test fails in idbnext. It is probably
+    // impossible to fix this. Unlike booleans and null, we can't remap objects cleanly,
+    // because:
+    //   a) they aren't constants, they could have anything in them
+    //   b) because someone else might want them to actually index on the inner parts
+    //      of the object (in this case, foo.foo)
+    it.skip('Orders different data types correctly, including objects', function () {
```
[the remark this answers is not in the document]
Comment (SCdF) [reply tests/find/test-suite-1/test.data-type-order.js]: I wrote some draft docs I threw against the indexeddb readme for now. We can move it to the right place when we move indexeddb there.

File: packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js [hunk 10]
```diff
@@ -2,73 +2,196 @@
...
+// when design documents trigger the addition or removal of native indexes.
+//
+// Given that Number.MAX_SAFE_INTEGER = 9007199254740991
+//
+// We can easily use the largest 2-3 digits and either allow:
+//  - 900 system migrations up to 2198/02/18
+//  - or 89 system migrations up to 5050/02/14
+//
+// This impl does the former. If this code still exists after 2198 someone send my
+// decendents a Spacebook message congratulating them on their impressive genes.
```
[the remark this answers is not in the document]
Comment (daleharvey) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/setup.js]: lol, agree with the tradeoff, 89 is a lot but a reachable number

**Comment by daleharvey** [comment]
pdb.get('_design/' + parts[0]).then(function (ddoc) {

then we dont need to double indent

**Comment by daleharvey** [comment]
Also we dont do `var a,b;`, just all new vars on new lines

File: packages/node_modules/pouchdb-adapter-indexeddb/src/find.js (hunk 2 above)
Comment (daleharvey) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: `codify` ?

**Comment by daleharvey** [comment]
So this is awesome, very impressed you managed to keep compatibility with the current `find` api I thought that was going to need to be changed, the schema upgrade mechanism is fairly similiar to a way that was proposed and looks solid. I am going to do a build and do a little more testing and reading to understand this but it looks great so far

I would like a little more time to review to fully understand, going to do a build and test it out a bit

**Comment by daleharvey** [comment]
> We still have more work to do before that (eg there is no migration path that I'm aware of). 
> I'm interested in what you think about instead exposing this as an optional adapter 
> so people can use it / test it if they want, e.g.:

We already do this, plus ship `pouchdb.next.js`

**Comment by daleharvey** [comment]
Ok, I have had a play around with this, I think long term we are going to want to look at replacing the abstract map reduce with a more specific and smaller peice of code that only needs to support `find`. I also want to start working on performance test infrastructure straight away but I am extremely happy to be getting this merged. Awesome work @SCdF, do you want to rebase this, fix any nits you fancy and we can get this landed?

**Comment by garrensmith** [comment]
Woohoo. Great work.


________________________________

File: packages/node_modules/pouchdb-adapter-indexeddb/src/find.js (hunk 2 above)
Comment (SCdF) [reply packages/node_modules/pouchdb-adapter-indexeddb/src/find.js]: I've gone with convert for now. I think that makes the most sense, its goal is to take a key that is valid for CouchDB and return one that is valid for IndexedDB.

We can always change it in a few if we take a later look and it's confusing.

**Comment by SCdF** [comment]
@daleharvey I've rebased and I'm happy for this to be merged in its current state :-)

**Comment by daleharvey** [comment]
woot, looks like the only test that is having issues is the PERF test, I dont think this code should change anything there so giving it a restart but that test rarely fails so maybe something up

**Comment by SCdF** [comment]
It finishes locally for me, but it definitely takes an age.

Definitely something to fix with an improved performance test infrastructure, either by having the framework output something occasionally to stop travis triggering, or running it on replicatable infrastructure etc

**Comment by daleharvey** [comment]
Cool, yeh definitely want to get in improving the performance test suite but in the meantime would be good to not having it broken by this commit, could you skip the test that takes too long or reduce its workload? Cheers

**Comment by SCdF** [comment]
@daleharvey OK this time the perf passed, but another one failed :-). If you could restart that, or chalk it up to flake, then hopefully we're ready to merge!
